src/core/query-builder.utils.ts

```diff
@@ -0,0 +1,76 @@
+import { QueryBuilderCustomOperation } from "smart-webcomponents-react";
+import { QueryBuilderOption } from "./types";
+import { ComputedDataFields } from "./query-builder.constants";
+
+/**
+ * Removes extra queries from computed fields so the query builder will read only one
+ * @param query actual query that has been sent
+ * @returns query that the query builder can read
+ */
+export function setQueryComputedFields(query?: string) {
+  const propertiesToExclude = Object.values(ComputedDataFields)
+  .flatMap(field => field.split(' || ').slice(1))
+  .join('|');
+  
+  const computedDataFieldsRegex = new RegExp(`\\s\\|\\| (${propertiesToExclude}) = *"[^"]*"`, "g");
+
+  return query?.replace(computedDataFieldsRegex, '');
+}
+
+/**
+ * The callback will replace the option's label with it's value
+ * @param options Object with value, label object properties, that hold the dropdown values
+ * @returns callback to be used by query builder when building the query
+ */
+export function expressionBuilderCallback(options: Record<string, QueryBuilderOption[]>) {
+  return function (this: QueryBuilderCustomOperation, fieldName: string, _operation: string, value: string) {
+    const buildExpression = (field: string, value: string) => {
+      const fieldOptions = options[fieldName];
+
+      if (fieldOptions?.length) {
+        const labelValue = fieldOptions.find(option => option.label === value);
+
+        if (labelValue) {
+          value = labelValue.value;
+        }
+      }
+
+      return this.expressionTemplate?.replace('{0}', field).replace('{1}', value);
+    };
+
+    if (fieldName.includes('||')) {
+      return `(${fieldName.split(' || ').map(prop => buildExpression(prop, value)).join(' || ')})`;
+    }
+
+    return buildExpression(fieldName, value);
+  };
+}
+
+/**
+ * The callback will replace the option's value with it's label
+ * @param options Object with value, label object properties, that hold the dropdown values
+ * @returns callback to be used by query builder when reading the query
+ */
+export function expressionReaderCallback(options: Record<string, QueryBuilderOption[]>) {
+  return function (_expression: string, [fieldName, value]: string[]) {
+    for (const computedFieldName of Object.values(ComputedDataFields)) {
+      if (computedFieldName.includes(fieldName)) {
+        fieldName = computedFieldName;
+        break;
+      }
+    }
+
+    const fieldOptions = options[fieldName];
+
+    if (fieldOptions?.length) {
+      const valueLabel = fieldOptions.find(option => option.value === value);
+
```

Comment:
nit: I like when the code gets a bit of space to breathe , however too many empty lines can also cause slight difficulty in reading.

My recommendation is to use empty lines only between different logical blocks. For instance the declaration of fieldOptions and the if check are part of the same logical block.